libslide/src/evaluator_rules/registry.rs

```diff
@@ -64,26 +69,43 @@ impl RuleSet {
         });
 
         let mut built_rules = Vec::with_capacity(num_rules);
+        let mut errors: Vec<Box<dyn Error>> = Vec::new();
         let bootstrapping_rules = Self::get_bootstrapping_rules();
-        for unbuilt_rule in self.rules.values() {
+        for unbuilt_rule in self.rules.values().chain(self.custom_rules.iter()) {
             match unbuilt_rule {
                 UnbuiltRule::S(s) => {
                     let pm = PatternMap::from_str(s);
+                    if let Some(err) = pm.validate() {
+                        errors.push(err.into());
+                        continue;
+                    }
+
                     let bootstrapped_pm = pm.bootstrap(&bootstrapping_rules);
                     built_rules.push(Rule::PatternMap(pm));
                     built_rules.push(Rule::PatternMap(bootstrapped_pm));
                 }
                 UnbuiltRule::F(f) => built_rules.push(Rule::from_fn(*f)),
             }
         }
-        built_rules
+
+        if !errors.is_empty() {
+            return Err(BuildRuleErrors { errors });
+        }
+
+        Ok(built_rules)
     }
 
-    /// Remove a rule from the rule set.
+    /// Remove a named rule from the rule set.
     pub fn remove(&mut self, rule: RuleName) {
         self.rules.remove(&rule);
     }
 
+    /// Insert a custom unbuilt rule into the rule set.
+    #[allow(unused)] // Used in testing. TODO: enable
+    fn insert_custom<T: Into<UnbuiltRule>>(&mut self, rule: T) {
+        self.custom_rules.push(rule.into());
+    }
+
```

Comment:
is this an error in linter because even functions used in testing do not come up with a warning otherwise like all of ours would.

Reply:
Yep it’s a linter (actually rust compiler) warning. Functions only used in testing but available publicly are considered dead code.